datetime: add new behavior for setting timezones

georgiy-belyanin opened the pull request:
This patch adds a fix making `datetime` handle `tz` and `tzoffset`
parameters properly.

Previously, changing a timezone altered the timestamp value but
preserved the time of day. Now you may enable new behavior making
specifying a timezone preserve the timestamp and update the represented
time of day.

This behavior could be achieved by setting a new compat option
`datetime_apply_timezone_action` to `'new'`. The option is going to be
enabled by default in 4.x.

Enabling new behavior involves the following.
1. Creating a datetime object with non-empty `tz`/`tzoffset` and
   `timestamp` values results with a time of day corresponding to
   the timestamp value plus the timezone offset. E.g if we create a
   datetime object with a timestamp corresponding to 9:12 o'clock in
   `+0000` timezone and set `tzoffset = -60` the resulting timestamp
   will represent 8:12 o'clock.
2. Setting a new `tz`/`tzoffset` affects the time of day represented by
   the datetime. E.g. if you update a datetime object of 11:12 o'clock
   having `tzoffset = 180` with `tzoffset = 120` the resulting datetime
   object will represent 10:12.

Closes #10363

Doc request tarantool/doc#4720

ligurio commented on test/app-tap/datetime.test.lua:
```diff
@@ -2819,27 +2819,27 @@ test:test("Time :set{} operations", function(test)
-    test:is(tostring(ts:set{ tzoffset = '+0800' }), '2020-11-09T06:12:23+0800',
+    test:is(tostring(ts:set{ tzoffset = '+0800' }), '2020-11-09T11:12:23+0800',
             'offset +0800' )
     -- timestamp 1630359071.125 is 2021-08-30T21:31:11.125Z
```
the comment should be updated too, right?

georgiy-belyanin replied:
I suppose the timestamp is in `+0000`
I've extended the comment
```diff
@@ -2823,7 +2823,8 @@ test:test("Time :set{} operations", function(test)
             'offset -0800' )
     test:is(tostring(ts:set{ tzoffset = '+0800' }), '2020-11-09T11:12:23+0800',
             'offset +0800' )
-    -- timestamp 1630359071.125 is 2021-08-30T21:31:11.125Z
+    -- timestamp 1630359071.125 is 2021-08-30T21:31:11.125Z+0000
+    -- or 2021-08-31T05:31:11.125Z+0800
```

ligurio reviewed (commented):
Georgiy, thanks for the patch! Please see my comments.

ligurio commented on changelogs/unreleased/gh-10363-datetime-handle-timezone.md:
```diff
@@ -0,0 +1,5 @@
+## bugfix/datetime
```
The comment should be closed after adding a doc request.

georgiy-belyanin replied:
Doc request tarantool/doc#4720

ligurio replied:
Usually we write doc request in commit message body (like this https://github.com/tarantool/tarantool/commit/4c302320c40f7ec2c81b10bf72287bb5779ab43b). Issue in doc/doc repository will be created automatically after merging.  However, a manually created issue is ok too.

georgiy-belyanin commented on test/app-tap/datetime.test.lua:
```diff
@@ -2631,18 +2631,18 @@ test:test("Parse with a custom format and format back to string", function(test)
             buf = '+0403',
             fmt = '%z',
-            dt = date.new({tzoffset = 243}),
+            dt = date.new({h = 0, m = 0, s = 0, tzoffset = 243}),
```
Thank you for noticing. Applied.

georgiy-belyanin commented on test/app-tap/datetime.test.lua:
```diff
@@ -2631,18 +2631,18 @@ test:test("Parse with a custom format and format back to string", function(test)
             buf = '-0403',
             fmt = '%z',
-            dt = date.new({tzoffset = -243}),
+            dt = date.new({h = 0, m = 0, s = 0, tzoffset = -243}),
```
Thank you for noticing. Applied.

georgiy-belyanin commented on test/app-tap/datetime.test.lua:
```diff
@@ -2631,18 +2631,18 @@ test:test("Parse with a custom format and format back to string", function(test)
             buf = 'MSK',
             fmt = '%Z',
-            dt = date.new({tzoffset = 180}),
+            dt = date.new({h = 0, m = 0, s = 0, tzoffset = 180}),
```
Thank you for noticing. Applied.

georgiy-belyanin commented on src/lua/datetime.lua:
```diff
@@ -584,6 +584,21 @@ local function datetime_new(obj)
+    local offset = obj.tzoffset
+    if offset ~= nil then
+        offset = get_timezone(offset, 'tzoffset')
+        -- at the moment the range of known timezones is UTC-12:00..UTC+14:00
```
This comment has been copy-pasted from the existing code though if you are OK I propose formatting it a bit:
```diff
@@ -588,7 +588,9 @@ local function datetime_new(obj)
     local offset = obj.tzoffset
     if offset ~= nil then
         offset = get_timezone(offset, 'tzoffset')
-        -- at the moment the range of known timezones is UTC-12:00..UTC+14:00
+        -- At the moment the range of known timezones is
+        -- UTC-12:00..UTC+14:00.
+        --
         -- https://en.wikipedia.org/wiki/List_of_UTC_time_offsets
         check_range(offset, -720, 840, 'tzoffset')
     end
```

georgiy-belyanin commented on src/lua/datetime.lua:
```diff
@@ -1149,13 +1166,23 @@ function datetime_set(self, obj)
         return self
     end
 
+    -- only timezone is changed
```
Applied.
```
@@ -1173,7 +1175,7 @@ function datetime_set(self, obj)
         return self
     end
 
-    -- only timezone is changed
+    -- Only timezone is changed.
     if not hms and not ymd then
         if tzname ~= nil then
```

georgiy-belyanin commented on src/lua/datetime.lua:
```diff
@@ -936,7 +940,20 @@ local function datetime_parse_from(str, obj)
     -- string.
     if date.tz == '' and date.tzoffset == 0 then
-        datetime_set(date, { tzoffset = tzoffset, tz = tzname })
+        local offset = nil
+
+        if tzoffset ~= nil then
+            offset = get_timezone(tzoffset, 'tzoffset')
+            check_range(offset, -720, 840, 'tzoffset')
+        end
+
+        if tzname ~= nil then
+            offset, date.tzindex = parse_tzname(date.epoch, tzname)
+        end
+
+        if offset ~= nil then
+            time_localize(date, offset)
+        end
```
This changes are already covered by the existing tests.
The change is needed since before the function used `datetime_set()` call for specifying a `tzoffset` and `tzname`.

But after the patch, `datetime_set` with a provided timezone not only changes the timezone but affect the hour/minute values. Though we only need to roughly set the timezone since hours/minutes are already properly parsed and calculated.

Thus I've copied the code from the `datetime_set` implementation here. It's already covered in the variety of tests on parsing.

```diff
@@ -936,7 +942,20 @@ local function datetime_parse_from(str, obj)
     -- Override timezone, if it was not specified in a parsed
     -- string.
     if date.tz == '' and date.tzoffset == 0 then
-        datetime_set(date, { tzoffset = tzoffset, tz = tzname })
+        local offset = nil
+
+        if tzoffset ~= nil then
+            offset = get_timezone(tzoffset, 'tzoffset')
+            check_range(offset, -720, 840, 'tzoffset')
+        end
+
+        if tzname ~= nil then
+            offset, date.tzindex = parse_tzname(date.epoch, tzname)
+        end
+
+        if offset ~= nil then
+            time_localize(date, offset)
+        end
     end
```

georgiy-belyanin committed:
datetime: refactor datetime_new timezone handling

This patch makes `datetime.new()` call internally determine the timezone
before calculating time since epoch when passing a timestamp as a
parameter. It's needed to fix handling timezones with timestamps to be
able to take the provided timezone into account when calculating time
since epoch.

Needed for #10363

NO_CHANGELOG=no visible changes
NO_TEST=no new behavior
NO_DOC=no new behavior

georgiy-belyanin committed:
datetime: add new behavior for setting timezones

This patch adds a fix making `datetime` handle `tz` and `tzoffset`
parameters properly.

Previously, changing a timezone altered the timestamp value but
preserved the time of day. Now you may enable new behavior making
specifying a timezone preserve the timestamp and update the represented
time of day.

This behavior could be achieved by setting a new compat option
`datetime_apply_timezone_action` to `'new'`. The option is going to be
enabled by default in 4.x.

Enabling new behavior involves the following.
1. Creating a datetime object with non-empty `tz`/`tzoffset` and
   `timestamp` values results with a time of day corresponding to
   the timestamp value plus the timezone offset. E.g if we create a
   datetime object with a timestamp corresponding to 9:12 o'clock in
   `+0000` timezone and set `tzoffset = -60` the resulting timestamp
   will represent 8:12 o'clock.
2. Setting a new `tz`/`tzoffset` affects the time of day represented by
   the datetime. E.g. if you update a datetime object of 11:12 o'clock
   having `tzoffset = 180` with `tzoffset = 120` the resulting datetime
   object will represent 10:12.

Closes #10363

NO_DOC=will be in tarantool/doc#4720

georgiy-belyanin commented on src/lua/datetime.lua:
```diff
@@ -640,12 +650,6 @@ local function datetime_new(obj)
-        offset, tzindex = parse_tzname(epoch_from_dt(dt), tzname)
-    end
-
     -- .hour, .minute, .second
```
Applied. Thank you.

georgiy-belyanin commented on changelogs/unreleased/gh-10363-datetime-handle-timezone.md:
```diff
@@ -0,0 +1,5 @@
+## bugfix/datetime
```
I've written a few tests similar to the examples:
```diff
--- a/test/app-luatest/datetime_test.lua
+++ b/test/app-luatest/datetime_test.lua
@@ -2098,3 +2098,55 @@ for supported_by, standard_cases in pairs(UNSUPPORTED_DATETIME_FORMATS) do
         end
     end
 end
+
+-- Test providing a timezone doesn't affect the timestamp
+-- value but instead it affects hours/minutes represented by
+-- datetime objects.
+--
+-- The scenario has been broken before gh-10363.
+pg.test_timestamp_with_tz_in_new_preserves_timestamp = function()
+    local now = dt.now()
+    local tzoffset = now.tzoffset
+
+    local now_from_timestamp_with_tzoffset = dt.new({
+        timestamp = now.timestamp,
+        tzoffset = tzoffset,
+    })
+    local now_from_timestamp_different_tzoffset = dt.new({
+        timestamp = now.timestamp,
+        tzoffset = tzoffset + 60,
+    })
+
+    -- Provided timestamp values aren't affected by tzoffsets.
+    t.assert_equals(now.timestamp, now_from_timestamp_with_tzoffset.timestamp)
+    t.assert_equals(now.timestamp, now_from_timestamp_different_tzoffset.timestamp)
+
+    -- Hours are updated correspondingly to the provided
+    -- tzoffsets.
+    t.assert_equals(now.hour, now_from_timestamp_with_tzoffset.hour)
+    t.assert_equals(now.hour + 1, now_from_timestamp_different_tzoffset.hour)
+end
+
+-- Test setting a timezone also changes the hour represented
+-- by a datetime object and a corresponding timestamp remains
+-- the same.
+--
+-- The scenario has been broken before gh-10363.
+pg.test_tz_updates_not_change_timestamps = function()
+    local now = dt.now()
+    local tzoffset = now.tzoffset
+
+    local now_with_tzoffset_plus_60 = dt.new({timestamp = now.timestamp})
+                                        :set({tzoffset = tzoffset + 60})
+    local now_with_tzoffset_minus_60 = dt.new({timestamp = now.timestamp})
+                                         :set({tzoffset = tzoffset - 60})
+
+    -- Timestamp values aren't affected by tzoffset changes.
+    t.assert_equals(now.timestamp, now_with_tzoffset_plus_60.timestamp)
+    t.assert_equals(now.timestamp, now_with_tzoffset_minus_60.timestamp)
+
+    -- Hours are updated correspondingly to the tzoffset
+    -- changes.
+    t.assert_equals(now.hour + 1, now_with_tzoffset_plus_60.hour)
+    t.assert_equals(now.hour - 1, now_with_tzoffset_minus_60.hour)
+end
```

And changed the changelog entry:
```diff
--- a/changelogs/unreleased/gh-10363-datetime-handle-timezone.md
+++ b/changelogs/unreleased/gh-10363-datetime-handle-timezone.md
@@ -1,5 +1,5 @@
 ## bugfix/datetime
 
-* `datetime` methods now handle timezones properly. They now behave like
-  offsets for timestamp values and only affect the way the date is
-  printed (gh-10363).
+* `datetime` methods now handle timezones properly. Now timezone changes
+  only affect the way timezone is formatted and don't affect the
+  represented timestamp (gh-10363).
```

What do you think? Is it better now?

ligurio replied:
Patch: **datetime: get timezone before HMS and YMD on new**

> datetime: get timezone before HMS and YMD on new

It is still not clear what do you mean: "on new" - what is it? "get timezone" - is it a public method "get' or not?
I propose to change to something like "datetime: refactor handling timezone in datetime_new"

>    This patch makes `datetime:new()` call internally determine the timezone

there is not such method in `datetime` module, but there is `datetime.new()` [^1]

>    before calculating time since epoch when passing a timestamp as a
>    parameter. It's needed to fix handling timezones with timestamps to be
>    able to take the provided timezone into account when calculating time
>    since epoch.

[^1]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/datetime/

georgiy-belyanin replied:
Applied your suggestions on the first patch.
The second commit message has been rewritten. Please, take a look. Thank you.

georgiy-belyanin commented:
@ligurio Please, take a look at the comments when you have time.

ligurio commented on changelogs/unreleased/gh-10363-datetime-handle-timezone.md:
```diff
@@ -0,0 +1,5 @@
+## bugfix/datetime
+
+* `datetime` methods now handle timezones properly. Now timezone changes
```
>  `datetime` methods now handle timezones properly.

too generic

Isn't it better?  "timestamp is not altered when it was set together with timezone"

ligurio commented on test/app-luatest/datetime_test.lua:
```diff
@@ -2098,3 +2098,56 @@ for supported_by, standard_cases in pairs(UNSUPPORTED_DATETIME_FORMATS) do
+-- the same.
+--
+-- The scenario has been broken before gh-10363.
+pg.test_tz_updates_not_change_timestamps = function()
```
It is a testcase that should be parameterized by `tzoffset`.

georgiy-belyanin replied:
Applied

ligurio commented on test/app-luatest/datetime_test.lua:
```diff
@@ -2098,3 +2098,56 @@ for supported_by, standard_cases in pairs(UNSUPPORTED_DATETIME_FORMATS) do
+-- datetime objects.
+--
+-- The scenario has been broken before gh-10363.
+pg.test_timestamp_with_tz_in_new_preserves_timestamp = function()
```
the same:  It is a testcase that should be parameterized by `tzoffset`.

georgiy-belyanin replied:
Applied

georgiy-belyanin commented on src/lua/datetime.lua:
```diff
@@ -936,7 +942,20 @@ local function datetime_parse_from(str, obj)
     -- Override timezone, if it was not specified in a parsed
     -- string.
     if date.tz == '' and date.tzoffset == 0 then
-        datetime_set(date, { tzoffset = tzoffset, tz = tzname })
```
I've implemented it by making the code adjust the timestamp if new datetime set timezone mode is enabled.

```diff
@@ -937,6 +979,13 @@ local function datetime_parse_from(str, obj)
     -- string.
     if date.tz == '' and date.tzoffset == 0 then
         datetime_set(date, { tzoffset = tzoffset, tz = tzname })
+
+        -- If working in a "preserve timestamp" mode the set
+        -- call changes represented time of day and it should be
+        -- adjusted (gh-10363).
+        if applying_tz_preserves_timestamp then
+            time_localize(date, date.tzoffset)
+        end
     end
 
     return date, len
```

ligurio commented on test/app-luatest/datetime_test.lua:
```diff
@@ -2098,3 +2099,151 @@ for supported_by, standard_cases in pairs(UNSUPPORTED_DATETIME_FORMATS) do
+-- Test new/old behavior of applying timezone to datetime objects
+-- (gh-10363).
+for _, case in ipairs(TIMEZONES) do
+    local function add_test_case(method, behavior, ...)
```
I would put this function outside the loop.

ligurio reviewed (approved):
Thanks for the patch! LGTM

tmr-g commented:
> It seems, this PR is better to merge after PR with refactoring of `new()` and `set()` #12410

#12410 is merged. The 'datetime: refactor datetime_new timezone handling' commit is not need anymore.

The new fix of `datetime.lua` for `new()` and `set()` with `timestamp` and `tz/tzoffset` params must be applied here: https://github.com/tarantool/tarantool/blob/a9f5c9086869b463d35f7a023e8bf6db0b8dd144/src/lua/datetime.lua#L655-L664

tmr-g commented:
It seems, we have such cases as a whole for `new()` and `set()`:

1. `ymd` and/or `hms` and/or `nsec` time units and `tz_new` -> given time units must be applied on existing time units of local time `tz_old` and it is used to make local time with `tz_new`. No new behavior.

```
datetime.new{hour = 3, tz = 'MSK'}:set{msec = 1, tzoffset=6*60}
---
- 1970-01-01T03:00:00.001+0600
...
```

2. `timestamp`, `tz` -> given `timestamp` is used for `epoch` for new behavior.

3. `tz` only
3.1 For `set()`: exists `epoch` (timestamp) is used for new behavior.
3.2 For `new()`: zero `epoch` with `tz` is used (as in 3.1) for new behavior. `datetime.new{tz=MSK} -> 1970.01.01 03:00:00 MSK`. It saves equivalence of constructor and set: `datetime.new{tz=x} == datetime.new{}:set{tz=x}`.

tmr-g commented on src/lua/datetime.lua:
```diff
@@ -1089,6 +1107,18 @@ function datetime_set(self, obj)
+    -- (gh-10363). Time units provided together with a timezone
+    -- are still applied to the wall clock time below.
+    if not ymd and not hms and nsec == nil and
+       tweaks.datetime_apply_timezone_preserves_timestamp then
```
May be we could add the 'has_nsec' result to `extract_obj_epoch_and_update_nsec()`? As we already have with `ymd`, `hms` flags.

The code will be more similar.

Feel free to ignore.

tmr-g commented on src/lua/datetime.lua:
```diff
@@ -991,6 +1002,13 @@ local function datetime_parse_from(str, obj)
     -- string.
     if date.tz == '' and date.tzoffset == 0 then
```
The optimal way to check 'tz is undefined' is `date.tzindex == 0`. Fix it also, please.

georgiy-belyanin replied:
Got it:

```diff
@@ -989,8 +1002,15 @@ local function datetime_parse_from(str, obj)
 
     -- Override timezone, if it was not specified in a parsed
     -- string.
-    if date.tz == '' and date.tzoffset == 0 then
+    if date.tzindex == 0 and date.tzoffset == 0 then
```

tmr-g commented on test/app-luatest/datetime_test.lua:
Please, add the vim-aware folding to highlight the new tests group.

```
-- {{{ title.
...
-- }}} title.
```

georgiy-belyanin replied:
Done! 

```lua
-- {{{ Compat option for applying timezones test (gh-10363).
```

tmr-g commented on test/app-luatest/datetime_test.lua:
The test cases implies whole-hour offsets, is it? Add the comment, please.

georgiy-belyanin replied:
```diff
+-- The listed timezones must not use DST at the moment and must
+-- have whole-hour offsets: the expected hour values below are
+-- calculated by adding the offsets converted to hours.
```

Yes, so here it is.

tmr-g commented on test/app-luatest/datetime_test.lua:
Can we get out of range [0..23] hour here?

georgiy-belyanin replied:
Theoretically yes. Just in case, I've added modulos here:

```diff
+    local old_expected = {
+        hour = (now.hour - now.tzoffset / 60) % 24,
+        timestamp = now.timestamp - case.tzoffset * 60,
+    }
+    local new_expected = {
+        hour = (now.hour + (case.tzoffset - now.tzoffset) / 60) % 24,
+        timestamp = now.timestamp,
+    }
```

So now it is in range for sure.

tmr-g commented on test/app-luatest/datetime_test.lua:
We must check somehow, what these standard offsets are 'correct' before tests. They can change in the future.

```
z = TIMEZONES[i]
assert(datetime.new{z.tzname}.tzoffset == z.tzoffset)
```

georgiy-belyanin replied:
I test it like this:

```diff
+    -- Ensure the hardcoded offset is still correct for the
+    -- timezone, it may change with a tzdata update.
+    assert(dt.new{timestamp = now.timestamp, tz = case.tzname}.tzoffset ==
+           case.tzoffset, ('outdated tzoffset for %s'):format(case.tzname))
```

tmr-g commented on test/app-luatest/datetime_test.lua:
One more case: `dt.new{}:set{timestamp = x, tz | tzoffset = y}`...

georgiy-belyanin replied:
Oh, you are right.

```diff
+    create_dt = function()
+        return dt.new{}:set{timestamp = now.timestamp,
+                            tzoffset = case.tzoffset}
+    end
+    add_test_case('set_timestamp_tzoffset', 'old', create_dt, old_expected)
+    add_test_case('set_timestamp_tzoffset', 'new', create_dt, new_expected)
+
+    create_dt = function()
+        return dt.new{}:set{timestamp = now.timestamp, tz = case.tzname}
+    end
+    add_test_case('set_timestamp_tzname', 'old', create_dt, old_expected)
+    add_test_case('set_timestamp_tzname', 'new', create_dt, new_expected)
+end
```

tmr-g reviewed (changes requested):
Thanks for the patch! Please, see my notes.

tmr-g commented on src/lua/datetime.lua:
```diff
@@ -585,15 +585,17 @@ end
 --
 -- 1. https://github.com/tarantool/tarantool/issues/12411
 local function extract_obj_epoch_and_update_nsec(obj, ymd, hms, nsec, from_set)
+    local has_nsec = nsec ~= nil
```
`nsec` may be defined below from fractional part of timestamp.
